torchrl/envs/libs/openspiel.py

```diff
@@ -0,0 +1,358 @@
+                device=self.device,
+            ),
+            current_player=DiscreteTensorSpec(
+                env.num_players(),
+                dtype=torch.int,
+                device=self.device,
+            ),
+            state=NonTensorSpec([]),
+            shape=self.batch_size,
+        )
+
+    def _set_seed(self, seed):
+        if seed is not None:
+            raise NotImplementedError("This environment has no seed.")
+
+    def current_player(self):
+        return self._env.current_player()
+
+    def _update_action_mask(self):
+        action_mask = torch.zeros(
+            self._env.num_distinct_actions(), device=self.device, dtype=torch.bool
+        )
+        action_mask[self._env.legal_actions()] = True
+        self.action_spec.update_mask(action_mask)
+
+    def _make_td_out(self, exclude_reward=False):
+        observation_shape = self._env.get_game().observation_tensor_shape()
+        observation = torch.stack(
+            [
+                self._to_tensor(self._env.observation_tensor(idx)).reshape(
+                    observation_shape
+                )
+                for idx in range(self._env.num_players())
+            ]
+        )
+        reward = self._to_tensor(self._env.returns())
+        done = torch.tensor(
+            self._env.is_terminal(), device=self.device, dtype=torch.bool
+        )
+        state = self._env.serialize()
+        current_player = torch.tensor(
+            self.current_player(), device=self.device, dtype=torch.int
+        )
+
+        tensordict_out = TensorDict(
+            source={
+                "observation": observation,
+                "done": done,
+                "terminated": done.clone(),
+                "state": state,
+                "current_player": current_player,
+            },
+            batch_size=self.batch_size,
+            device=self.device,
+        )
+
+        if not exclude_reward:
+            tensordict_out["reward"] = reward
+
+        return tensordict_out
+
+    def _step(self, tensordict: TensorDictBase) -> TensorDictBase:
+        action = tensordict.get("action")
+
+        if not self._categorical_action_encoding:
+            # TODO: To support this, need to convert the one-hot to an index value
+            raise NotImplementedError
+
+        self._env.apply_action(action)
+        self._update_action_mask()
+        return self._make_td_out()
+
+    def _to_tensor(self, value):
+        return torch.tensor(value, device=self.device, dtype=torch.float32)
+
+    def _reset(
+        self, tensordict: Optional[TensorDictBase] = None, **kwargs
+    ) -> TensorDictBase:
+        game = self._env.get_game()
+
+        if tensordict is not None and "state" in tensordict:
+            new_env = game.deserialize_state(tensordict["state"])
+        else:
+            new_env = game.new_initial_state()
+
+        self._env = new_env
+        self._update_action_mask()
+        return self._make_td_out(exclude_reward=True)
```

Comment:
interesting, so openspiel has a reward at reset time?

Reply:
Yes, but it's just set to zero